File: go-version/structures/stack.go
package structures

import (
	"encoding/json"
	"os"
	"strconv"
)

// StackNode - узел стека
type StackNode struct {
	data string
	next *StackNode
}

// Stack - стек (LIFO)
type Stack struct {
	top *StackNode
}

// NewStack создает новый стек
func NewStack() *Stack {
	return &Stack{top: nil}
}

// Push добавляет элемент в стек
func (s *Stack) Push(value string) string {
	newNode := &StackNode{data: value, next: s.top}
	s.top = newNode
	return value
}

// Pop удаляет и возвращает верхний элемент
func (s *Stack) Pop() (string, bool) {
	if s.top == nil {
		return "", false
	}
	value := s.top.data
	s.top = s.top.next
	return value, true
}

// IsEmpty проверяет, пуст ли стек
func (s *Stack) IsEmpty() bool {
	return s.top == nil
}

// Values возвращает все элементы стека (от вершины к дну)
func (s *Stack) Values() []string {
	var values []string
	current := s.top
	for current != nil {
		values = append(values, current.data)
		current = current.next
	}
	return values
}

// Print выводит стек
func (s *Stack) Print(name string) string {
	result := "Stack [" + name + "] (top to bottom):\n"
	current := s.top
	index := 0
	for current != nil {
		result += "  [" + strconv.Itoa(index) + "] | " + current.data + "\n"
		current = current.next
		index++
	}
	if index == 0 {
		result += "  (empty)\n"
	}
	return result
}

// Serialize сохраняет стек в JSON файл
func (s *Stack) Serialize(filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	encoder := json.NewEncoder(file)
	encoder.SetIndent("", "    ")
	return encoder.Encode(s.Values())
}

// Deserialize загружает стек из JSON файла
func (s *Stack) Deserialize(filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	decoder := json.NewDecoder(file)
	var values []string
	if err := decoder.Decode(&values); err != nil {
		return err
	}
	s.top = nil
	// Iterate in reverse to restore stack order
	for i := len(values) - 1; i >= 0; i-- {
		s.Push(values[i])
	}
	return nil
}

File: src/Stack.cpp
#include "../include/Stack.h"
#include <nlohmann/json.hpp>

using json = nlohmann::json;

Stack::Stack() : topNode(nullptr), count(0) {}

Stack::~Stack() {
    clear();
}

void Stack::push(const std::string& value) {
    StackNode* newNode = new StackNode(value);
    newNode->next = topNode;
    topNode = newNode;
    count++;
}

void Stack::pop() {
    if (isEmpty()) return;
    StackNode* temp = topNode;
    topNode = topNode->next;
    delete temp;
    count--;
}

std::string Stack::top() const {
    if (isEmpty()) {
        throw std::runtime_error("Stack is empty");
    }
    return topNode->data;
}

bool Stack::isEmpty() const {
    return topNode == nullptr;
}

int Stack::size() const {
    return count;
}

void Stack::print() const {
    StackNode* temp = topNode;
    while (temp) {
        std::cout << temp->data << " -> ";
        temp = temp->next;
    }
    std::cout << "NULL" << std::endl;
}

void Stack::clear() {
    while (!isEmpty()) {
        pop();
    }
}

void Stack::serialize(const std::string& filename) const {
    json j = json::array();
    StackNode* temp = topNode;
    while (temp) {
        j.push_back(temp->data);
        temp = temp->next;
    }
    std::ofstream file(filename);
    if (!file.is_open()) {
        throw std::runtime_error("Could not open file for writing");
    }
    file << j.dump(4);
    file.close();
}

void Stack::deserialize(const std::string& filename) {
    std::ifstream file(filename);
    if (!file.is_open()) {
        throw std::runtime_error("Could not open file for reading");
    }
    json j;
    file >> j;
    clear();
    // Iterate in reverse to push bottom elements first, restoring the stack order
    for (auto it = j.rbegin(); it != j.rend(); ++it) {
        push(it->get<std::string>());
    }
    file.close();
}

void Stack::serializeBinary(const std::string& filename) const {
    std::ofstream file(filename, std::ios::binary);
    StackNode* temp = topNode;
    while (temp) {
        int len = temp->data.length();
        file.write(reinterpret_cast<const char*>(&len), sizeof(len));
        file.write(temp->data.c_str(), len);
        temp = temp->next;
    }
    file.close();
}

void Stack::deserializeBinary(const std::string& filename) {
    clear();
    std::ifstream file(filename, std::ios::binary);
    Stack tempStack;
    while (file.peek() != EOF) {
        int len;
        file.read(reinterpret_cast<char*>(&len), sizeof(len));
        char* buffer = new char[len + 1];
        file.read(buffer, len);
        buffer[len] = '\0';
        tempStack.push(std::string(buffer));
        delete[] buffer;
    }
    file.close();
    
    while (!tempStack.isEmpty()) {
        push(tempStack.top());
        tempStack.pop();
    }
}
